lib/IDE/Refactoring.cpp

```diff
@@ -2601,6 +2601,46 @@ bool RefactoringActionLocalizeString::performChange() {
   EditConsumer.insertAfter(SM, Target->getEndLoc(), ", comment: \"\")");
   return false;
 }
+    
+bool RefactoringActionDocCommentBoilerplate::isApplicable(ResolvedCursorInfo Tok, DiagnosticEngine &Diag) {
+#warning Implement here
+    if (Tok.Kind != CursorInfoKind::ValueRef)
+        return false;
+    
+    return Tok.ValueD->getKind() == DeclKind::Func;
+}
+    
+bool RefactoringActionDocCommentBoilerplate::performChange() {
+#warning Implement here
+    if(CursorInfo.ValueD->getKind() == DeclKind::Func) {
+        swift::SourceLoc startLocation = CursorInfo.ValueD->getStartLoc();
+        
+        if(isa<FuncDecl>(CursorInfo.ValueD)) {
+            SmallVector<StringRef, 5> outputLines;
+            EditConsumer.accept(SM, startLocation, "/**\n<#Function Summary#>\n");
+            
+            FuncDecl *funcDecl = (FuncDecl *)CursorInfo.ValueD;
+            for (auto paramLists : funcDecl->getParameterLists()) {
+                auto paramDecls = paramLists->getArray();
+                for(auto paramDecl : paramDecls) {
+                    const char *parameterName = paramDecl->getName().get();
+                    
+                    EditConsumer.accept(SM, startLocation, "  - Parameter ");
```

Comment:
Can you make this generate 
````
- Parameters:
  - foo: ...
  - bar: ...
```

If there is more than one parameter?

Reply:
What is the right way to get the number of parameters?
```
for (auto paramLists : funcDecl->getParameterLists()) {
        auto paramDecls = paramLists->getArray();
        for(auto paramDecl : paramDecls) {
            const char *parameterName = paramDecl->getName().get();
            
            EditConsumer.accept(SM, startLocation, "///  - Parameter ");
            EditConsumer.accept(SM, startLocation, parameterName);
            EditConsumer.accept(SM, startLocation, ": <#Parameter ");
            EditConsumer.accept(SM, startLocation, parameterName);
            EditConsumer.accept(SM, startLocation, "#>\n");
        }
    }
```
I have nested loops here and not sure if I should check ```funcDecl->getParameterLists()``` or ```paramLists->getArray()```

Reply:
You can always flatten the multiple arrays into one full array. We removed the syntax for multiple parameter lists in one function, right now they’re vestigial and only used in subscripts (if I remember correctly).

Reply:
So would it be safe to do
```
for(auto paramDecl : funcDecl->getParameterLists()->getArray()) {
    ...
}
```
?

Reply:
You’d have to build an intermediate array of all the parameters in all the lists, and then loop over those.

Reply:
Ahhh I see ... On it! Sorry that I need some hand holding here!